src/types.py

```diff
@@ -930,6 +943,117 @@ def ctype(ctype: CType, typemap: TypeMap, typepool: TypePool) -> "Type":
             return Type(TypeData(kind=TypeData.K_INT, size_bits=size_bits, sign=sign))
         static_assert_unreachable(real_ctype)
 
+    @staticmethod
+    def demangled_symbol(sym: CxxSymbol) -> "Type":
+        return Type.demangled_type(sym_type=sym.type, sym_name=sym.name)
+
+    @staticmethod
+    def demangled_type(sym_type: CxxType, sym_name: Optional[CxxTerm] = None) -> "Type":
+        # TODO: This function may need to depend on the ABI or compiler to generate the types
+        # of class member functions. It may need to be moved out of types.py, or some of this
+        # logic could be moved into FunctionSignature?
+        final_name = None
+        if sym_name is not None:
+            assert sym_name.kind == CxxTerm.Kind.QUALIFIED
+            assert sym_name.qualified_name is not None
+            assert len(sym_name.qualified_name) >= 1
+            final_name = str(sym_name.qualified_name[-1]).split("@")[-1]
+
+        type = None
+        for term in sym_type.terms[::-1]:
+            if type is None and term.kind in CxxTerm.NONTERMINATING_KINDS:
+                type = Type.any()
+            if term.kind == CxxTerm.Kind.CONST:
+                pass
+            elif term.kind == CxxTerm.Kind.UNSIGNED:
+                unsigned = True
+            elif term.kind in (CxxTerm.Kind.POINTER, CxxTerm.Kind.REFERENCE):
+                type = Type.ptr(type).weaken_void_ptr()
```

Comment:
The idea is to replace `Type.ptr(Type.void())` with `Type.ptr(None)`. (`void *`s get mangled into `Pv` which is parsed into `[POINTER, VOID]`)

It might be a good idea to add a special case to `Type.unify()` and/or refactor how the ctypes parsing behaves -- but I think that's a refactor that'd be better to do separately?

Reply:
Hm, my comment here was a bit unclear: I was more interested in the why than the what. I thought we only did this void* weakening when using void pointers in certain circumstances, not when creating types from them in general. E.g. when parsing C contexts I don't think we do this weakening?

I feel like I remember a discussion about unifying void with anything in unify() and deciding against it.

(Totally fine to leave this here for now on the basis of "it gives better results", naturally, it just struck me as weird)